lib/coq/Parser.vy

```diff
@@ -0,0 +1,76 @@
+%{
+
+Require Import String.
+Require Import List.
+
+Module Cst.
+Inductive obj :=
+  | TType : nat -> obj
+  | Nat : obj
+  | Zero : obj
+  | Succ : obj -> obj
+  | App : obj -> obj -> obj
+  | Fun : string -> obj -> obj -> obj
+  | Pi : string -> obj -> obj -> obj
+  | Var : string -> obj.
+End Cst.
+
+%}
+
+%token <string> VAR
+%token <nat> INT
+%token ZERO LAMBDA PI SUCC NAT TYPE (* keywords *)
+%token LPAREN RPAREN DOT COLON EOF (* delimiters *)
+
+%start <Cst.obj> prog
+%type <Cst.obj> obj app_obj simpl_obj
+%type <string * Cst.obj> args_obj
+%type <list (string * Cst.obj)> args_list rev_args_list
+
+%%
+
+prog:
+  | obj EOF { $1 }
+
+obj:
+  (* An object is a lambda / pi, *)
+  | LAMBDA args_list DOT obj {
+      List.fold_left (fun acc arg => Cst.Fun (fst arg) (snd arg) acc) $2 $4 
+  }
+  | PI args_list DOT obj {
+      List.fold_left (fun acc arg => Cst.Pi (fst arg) (snd arg) acc) $2 $4
+  }
+  (* or an application (a "variable" is an application with no arguments) *)
+  (* see https://github.com/utgwkk/lambda-chama/blob/master/parser.mly *)
+  | app_obj { $1 }
+
+(* To avoid quadratic time in parsing the list of arguments, we must
+    prepend the arguments and reverse it at the end. *)
+args_list:
+  | rev_args_list { List.rev $1 }
+
+rev_args_list:
+  (* Nonempty list of arguments *)
+  | rev_args_list args_obj { $2 :: $1 }
+  | args_obj { [$1] }
+
+(* (x : A) *)
+args_obj:
+  | LPAREN VAR COLON obj RPAREN { ($2, $4) }
+
+(* M N *)
+app_obj:
+  (* simpl_obj prevents conflict by associativity *)
+  | app_obj simpl_obj { Cst.App $1 $2 }
+  | simpl_obj { $1 }
+
+(* Either a primitive object like "Type 5" / "Zero"
+  or parentheses around a complex object *)
+simpl_obj:
```

Comment:
all cases other than VAR and parentheses can be placed in the obj rule, as none of the rest can be used as a function.